### ISSUE TITLE
Nondeterministic results of LSTM

### ISSUE THREAD

**Issue by yuanx749** [pytorch/pytorch#62096]
## ❓ Questions and Help

When using dropout in a multilayer LSTM trained on GPU, the weights are different across training. I have reset the same seed before training, so the dataloader has the same behaviour and the model has the same initialization. Note that the results are the same across different runs of the whole script, but different across call of the training routine in the same run of the script.

Say if I use Jupyter notebook, and I repetitively run a cell that resets the seed and trains, the results will be different. What else can I do if want the same result?

## To Reproduce

Run the script below. `CUBLAS_WORKSPACE_CONFIG=:16:8 python train.py`

```python
import random
import copy
import numpy as np
import torch
import torch.nn as nn
from torch.utils.data import DataLoader
from torch.utils.data.dataset import Dataset


class NN(nn.Module):
    def __init__(self):
        super().__init__()
        self.lstm = nn.LSTM(input_size=6, hidden_size=32, num_layers=2, batch_first=True, dropout=0.1)
        self.linear = nn.Linear(32, 2)
    
    def forward(self, inputs):
        outputs, _ = self.lstm(inputs)
        outputs = torch.mean(outputs, 1)
        outputs = self.linear(outputs)
        return outputs

class Data(Dataset):
    def __init__(self):
        super().__init__()
        self.data = np.random.rand(300).astype(np.float32).reshape(10, 5, 6)
        self.labels = np.random.choice(2, 10).astype(np.int64)

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        return self.data[idx], self.labels[idx]


def train(model, dataloader, criterion, optimizer):
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    for _, batch in enumerate(dataloader):
        inputs, labels = batch
        inputs = inputs.to(device)
        labels = labels.to(device)
        outputs = model(inputs)
        loss = criterion(outputs, labels)
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
    return

def run_experiment(epochs=2):
    
    train_dataset = Data()
    train_loader = DataLoader(train_dataset, batch_size=1, shuffle=True)
    
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    model = NN().to(device)
    
    optimizer = torch.optim.Adam(model.parameters())
    criterion = nn.CrossEntropyLoss()
    
    for _ in range(epochs):
        train(model, train_loader, criterion, optimizer)    
    
    return copy.deepcopy(model)

def compare_models(model_1, model_2):
    models_differ = 0
    for kv_1, kv_2 in zip(model_1.state_dict().items(), model_2.state_dict().items()):
        assert kv_1[0] == kv_2[0]
        if not torch.equal(kv_1[1], kv_2[1]):
            models_differ += 1
            print('Mismtach found: ', kv_1[0])
    if models_differ == 0:
        print('Models match.')

def set_seed(seed):
    torch.manual_seed(seed)
    random.seed(seed)
    np.random.seed(seed)
    torch.backends.cudnn.benchmark = False
    torch.backends.cudnn.deterministic = True

if __name__ == '__main__':

    SEED = 42    
    
    set_seed(SEED)
    m1 = run_experiment()
    
    set_seed(SEED)
    m2 = run_experiment()
    
    compare_models(m1, m2)

```

## Environment

PyTorch version: 1.9.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.12.0
Libc version: glibc-2.26

Python version: 3.7.11 (default, Jul  3 2021, 18:01:19)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
Is CUDA available: True
CUDA runtime version: 11.0.221
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 460.32.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.9.0+cu102
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.10.0
[pip3] torchvision==0.10.0+cu102
[conda] Could not collect

**Comment by jbschlosser** [comment]
Hey @yuanx749! I suggest posting on the [PyTorch forums](https://discuss.pytorch.org/), as you are more likely to be assisted there. The issues here are reserved for PyTorch feature requests / bug fixes, and it appears that deterministic runs are working correctly at the whole-script level.
